File: 279.完全平方数.py
#
# @lc app=leetcode.cn id=279 lang=python3
#
# [279] 完全平方数
#
from collections import deque

# @lc code=start
class Solution:
    def numSquares(self, n: int) -> int:
        """Dynamic programming approach"""
        square_nums = [i**2 for i in range(0, int(n**0.5)+1)]
        
        # numSquares(n) = min(numSquares(n-k) - 1 for k in squareNumbers)
        dp = [float('inf')] * (n+1)
        dp[0] = 0   # bottom case
        
        for i in range(1, n+1):
            for square in square_nums:
                if i < square:
                    break
                dp[i] = min(dp[i], dp[i-square] + 1)
        
        return dp[-1]

    
    def numSquares(self, n: int) -> int:
        """BFS approach, much faster than DP in python3.
        Recall that BFS can give the shortest path because
        it performs level traversal in the search tree."""
        if not n: 
            return 0
        queue = deque([n])
        step = 0
        visited = set()
        while queue:
            step += 1
            l = len(queue)
            for _ in range(l):
                tmp = queue.pop()
                for i in range(1, int(tmp**0.5)+1):     # square numbers for `tmp`
                    x = tmp - i**2
                    if not x:
                        return step
                    if x not in visited:
                        queue.appendleft(x)
                        visited.add(x)
        return step


# @lc code=end

File: 279.完全平方数.java
/*
 * @lc app=leetcode.cn id=279 lang=java
 *
 * [279] 完全平方数
 */

// @lc code=start
class Solution {
    /**
     * 问题的上界：
     *  1. 四平方数定理：正整数n可表示为4个正整数的平方和a^2 + b^2 + c^2 + d^2
     * 
     *  2. 三平方数定理：若正整数n不等于4^k * (8*m + 7)的形式，k和m为正整数，则n可表示为3个正整数的平方和
     * 
     */
    public int numSquares(int n) {
        if (isSq(n))  return 1;

        while (n % 4 == 0) {
            n /= 4;
        }
        if (n % 8 == 7)  return 4;  // 不符合三平方数条件

        // 检查是否能分解为两个平方数的和，只能枚举
        for (int i = 0; i * i <= n; ++i) {
            if (isSq(n - i * i))  return 2;
        }

        return 3;   // 找不到两个平方数，只能为三个
    }

    /** 检查一个数是否为平方数 */
    private boolean isSq(int n) {
        int sq = (int) Math.sqrt(n);
        return sq * sq == n;
    }
}
// @lc code=end

